x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizer.java

```diff
@@ -804,4 +824,49 @@ protected LogicalPlan rule(Limit plan) {
             return p;
         }
     }
+
+    public static class ReplaceRegexMatch extends OptimizerRules.ReplaceRegexMatch {
+
+        protected Expression regexToEquals(RegexMatch<?> regexMatch, Literal literal) {
+            return new Equals(regexMatch.source(), regexMatch.field(), literal);
+        }
+    }
+
+    private static class ReplaceAttributesWithExact extends OptimizerRules.OptimizerRule<LogicalPlan> {
+
+        @Override
+        protected LogicalPlan rule(LogicalPlan plan) {
+            if (plan instanceof LocalRelation || plan instanceof Drop) {
+                return plan;
+            } else if (plan instanceof Project proj) {
+                return plan.transformExpressionsOnlyUp(FieldAttribute.class, x -> toExactAlias(x, proj.projections()));
+            } else {
+                return plan.transformExpressionsOnly(FieldAttribute.class, ReplaceAttributesWithExact::toExact);
+            }
+        }
+
+        private NamedExpression toExactAlias(FieldAttribute e, List<? extends NamedExpression> projections) {
+
+            if (e.getExactInfo().hasExact() && e.exactAttribute() != e) {
+                FieldAttribute calculatedExact = toExact(e);
+
+                // avoid using multiple IDs just to load the same field twice
+                NamedExpression exact = projections.stream()
+                    .filter(x -> x.name().equals(calculatedExact.name()))
+                    .map(NamedExpression.class::cast)
+                    .findFirst()
+                    .orElse(calculatedExact);
+
+                return new Alias(e.source(), e.name(), exact);
+            }
+            return e;
+        }
+
+        private static FieldAttribute toExact(FieldAttribute fa) {
+            if (fa.getExactInfo().hasExact() && fa.exactAttribute() != fa) {
+                return fa.exactAttribute();
+            }
+            return fa;
```

Comment:
No need to add new aliases and fields - simply replace the FieldAttribute in place, with the same underlying NameId but set the EsField to its exact counterpart. That is the attribute remains essentially the same, it's just the backing field that has changed.